src/main/java/com/hubspot/jinjava/lib/tag/eager/EagerSetTag.java

```diff
@@ -0,0 +1,133 @@
+package com.hubspot.jinjava.lib.tag.eager;
+
+import com.hubspot.jinjava.interpret.Context;
+import com.hubspot.jinjava.interpret.DeferredValueException;
+import com.hubspot.jinjava.interpret.JinjavaInterpreter;
+import com.hubspot.jinjava.interpret.TemplateSyntaxException;
+import com.hubspot.jinjava.lib.tag.DoTag;
+import com.hubspot.jinjava.lib.tag.SetTag;
+import com.hubspot.jinjava.tree.parse.TagToken;
+import com.hubspot.jinjava.util.ChunkResolver;
+import java.util.Arrays;
+import java.util.List;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+public class EagerSetTag extends EagerStateChangingTag<SetTag> {
+
+  public EagerSetTag() {
+    super(new SetTag());
+  }
+
+  public EagerSetTag(SetTag setTag) {
+    super(setTag);
+  }
+
+  @Override
+  public String getEagerTagImage(TagToken tagToken, JinjavaInterpreter interpreter) {
+    if (!tagToken.getHelpers().contains("=")) {
+      throw new TemplateSyntaxException(
+        interpreter,
+        tagToken.getImage(),
+        "Tag 'set' expects an assignment expression with '=', but was: " +
+        tagToken.getHelpers()
+      );
+    }
+
+    int eqPos = tagToken.getHelpers().indexOf('=');
+    String variables = tagToken.getHelpers().substring(0, eqPos).trim();
+
+    String expression = tagToken.getHelpers().substring(eqPos + 1);
+    if (interpreter.getContext().containsKey(Context.IMPORT_RESOURCE_ALIAS)) {
+      return interpreter.render(
+        convertSetToUpdate(variables, expression, tagToken, interpreter)
+      );
+    }
+    ChunkResolver chunkResolver = new ChunkResolver(expression, tagToken, interpreter);
+    EagerStringResult resolvedExpression = executeInChildContext(
+      eagerInterpreter -> chunkResolver.resolveChunks(),
+      interpreter,
+      true
+    );
+    StringJoiner joiner = new StringJoiner(" ");
+    joiner
+      .add(tagToken.getSymbols().getExpressionStartWithTag())
+      .add(tagToken.getTagName())
+      .add(variables)
+      .add("=")
+      .add(resolvedExpression.getResult())
+      .add(tagToken.getSymbols().getExpressionEndWithTag());
+    StringBuilder prefixToPreserveState = new StringBuilder(
+      interpreter.getContext().isProtectedMode()
+        ? resolvedExpression.getPrefixToPreserveState()
+        : ""
+    );
+    String[] varTokens = variables.split(",");
+
+    if (
+      chunkResolver.getDeferredWords().isEmpty() &&
+      !interpreter.getContext().isProtectedMode()
```

Comment:
did we decide on a different name for `isProtectedMode`? Every time I see it, I can't remember what it protects from.

Reply:
I have not yet come up with a better name, and I'm aware that it's not exactly clear what it does by the name.

I have defined "protected mode" as a mode where the surrounding context cannot be modified such that the state of the context before entering "protected mode" will be the same as after exiting it. Any changes will result in the values attempted to be modified becoming `DeferredValue`s.

This mode is commonly seen in IfTags and ForTags when the children are being rendered but their results need to be protected such that when the if branch/for loop is finally executed normally, the output will be correct.

Something like `isDeferredExecutionMode` might be more understandable as the code being evaluated is being partially deferred, knowing that it's final execution is being deferred

Reply:
`isDeferredExecutionMode` sounds good to me.